src/commons/crypto/signing/dispatch/signerrouter.rs

```diff
@@ -159,159 +141,52 @@ impl SignerRouter {
-        // signer to use to create a new key depends on whether the key is one-off or not and whether or not it is
-        // being created for a key roll.
-
-        // TODO: Once it becomes possible to configure how an HSM is used by Krill we need to decide what the
-        // defaults should be and what should be configurable or not concerning HSM usage, and to document why, if
-        // permitted, it is acceptable to use local keys, signing & random number genration instead of the more
-        // secure HSM based alternatives (if available).
-        let signer_mapper = Arc::new(SignerMapper::build(work_dir)?);
-
-        // We don't know the signer handles yet. The signer implementations have to work out their own signer handle
-        // when they are ready. Signers are moved from the signer collection to the map once while running when asked
-        // they by that point have determined their own handle.
-        let signers_by_handle = RwLock::new(HashMap::new());
-
-        let roles = Self::build_signers(work_dir, signer_mapper.clone(), alternate_config)?;
-
-        // Having the same signer multiple times in this vector is less efficient but the impact is negligible and it
-        // doesn't break anything if there are duplicates.
-        let mut unique_signers = vec![
-            roles.default_signer.clone(),
-            roles.one_off_signer.clone(),
-            roles.rand_fallback_signer.clone(),
-        ];
-
-        unique_signers.sort_by_key(|k| k.get_name().to_string());
-        unique_signers.dedup_by_key(|k| k.get_name().to_string());
-
-        let unique_signers = RwLock::new(unique_signers);
-
-        // TODO: Once we can configure ourselves from the configuration file, we also need to create any signers that
-        // are defined in the config but which don't have an active role, i.e. only exist to work with keys created by
-        // a still active signer but not one that we create new keys with.
+    pub(crate) fn build(signer_mapper: Arc<SignerMapper>, mut signers: Vec<SignerProvider>) -> KrillResult<Self> {
+        // Keep a mapping of signer mapper handle to signer provider. Fill it in as and when signers become ready at
+        // which point their signer mapper handle will be known.
+        let active_signers = RwLock::new(HashMap::new());
+
+        // One and only one signer should be the default. The default signer is used for operations that don't concern
+        // an existing key, i.e. key creation, one-off signing and random number generation. The latter two may be
+        // delegated below to other signers instead of the default signer.
+        // Create the signers
+        let mut default_signer: Option<Arc<SignerProvider>> = None;
+        let mut one_off_signer: Option<Arc<SignerProvider>> = None;
+        let mut rand_falback_signer: Option<Arc<SignerProvider>> = None;
+        let mut all_signers = Vec::new();
+
+        for signer in signers.drain(..) {
```

Comment:
as mentioned elsewhere I wonder whether we should do this in `Config::verify`. That could keep things simpler and more focussed here.

Reply:
see other comment, let's keep it here for now